client.go

```diff
@@ -552,9 +570,9 @@ func (c *Client) moveToClosed() {
 		})
 	}
 
-	c.mu.RLock()
+	c.mu.Lock()
 	disconnectedCh := c.disconnectedCh
-	c.mu.RUnlock()
+	c.mu.Unlock()
 	// At this point connection close was issued, so we wait until the reader goroutine
 	// finishes its work, after that it's safe to close the callback queue.
 	if disconnectedCh != nil {
```

Comment:
Why is this such a pattern in this codebase? What's wrong with changing to the following? Also, can't the `c.disconnectedCh` not be nil here, but be nil on the next line? Should there be some sort of lock around that functionality?
```suggestion
	// At this point connection close was issued, so we wait until the reader goroutine
	// finishes its work, after that it's safe to close the callback queue.
	if c.disconnectedCh != nil {
```